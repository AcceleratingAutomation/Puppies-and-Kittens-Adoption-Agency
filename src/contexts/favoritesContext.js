import React, { createContext, useReducer } from 'react';

export const initialState = {
  adopters: [],
  fosters: [],
  rescues: [],
  veterinarians: [],
  users: [],
  rescueDetails: null,
  favorites: [],
  loading: true,
  openDialog: false
};

export function reducer(state, action) {
  switch (action.type) {
    case 'setAdopters':
      return { ...state, adopters: action.value, loading: false };
    case 'setFosters':
      return { ...state, fosters: action.value, loading: false };
    case 'setRescues':
      return { ...state, rescues: action.value, loading: false };
    case 'setUsers':
      return { ...state, users: action.value, loading: false };
    case 'setVeterinarians':
      return { ...state, veterinarians: action.value, loading: false };
    case 'setFavorites':
      return { ...state, favorites: action.value, loading: false };
    case 'addToFavorites':
      return {
        ...state,
        rescues: state.rescues.map(rescue =>
          rescue.id === action.id ? { ...rescue, isFavorite: true } : rescue
        ),
      };
    case 'removeFromFavorites':
      return {
        ...state,
        rescues: state.rescues.map(rescue =>
          rescue.id === action.id ? { ...rescue, isFavorite: false } : rescue
        ),
        favorites: state.favorites.filter(favoriteId => favoriteId !== action.id)
      };
    case 'setLoading':
      return { ...state, loading: action.value };
    default:
      throw new Error();
  }
}

export const FavoritesContext = createContext();

export const FavoritesProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <FavoritesContext.Provider value={{ state, dispatch }}>
      {children}
    </FavoritesContext.Provider>
  );
};