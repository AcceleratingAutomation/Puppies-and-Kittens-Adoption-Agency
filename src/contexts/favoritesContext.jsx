import React, { createContext, useReducer, useMemo } from "react";
import PropTypes from "prop-types";

export const initialState = {
  adopters: [],
  favorites: [],
  fosters: [],
  loading: true,
  openDialog: false,
  rescueDetails: null,
  rescues: [],
  veterinarians: [],
  users: [],
};

export function reducer(state, action) {
  switch (action.type) {
    case "setAdopters":
      return { ...state, adopters: action.value, loading: false };
    case "setFosters":
      return { ...state, fosters: action.value, loading: false };
    case "setRescues":
      return { ...state, rescues: action.value, loading: false };
    case "setRescueDetails":
      return { ...state, rescueDetails: action.value, loading: false };
    case "setUsers":
      return { ...state, users: action.value, loading: false };
    case "setVeterinarians":
      return { ...state, veterinarians: action.value, loading: false };
    case "setFavorites":
      return { ...state, favorites: action.value, loading: false };
    case "addToFavorites":
      return {
        ...state,
        rescues: state.rescues.map((rescue) =>
          rescue.id === action.id ? { ...rescue, isFavorite: true } : rescue,
        ),
      };
    case "removeFromFavorites":
      return {
        ...state,
        rescues: state.rescues.map((rescue) =>
          rescue.id === action.id ? { ...rescue, isFavorite: false } : rescue,
        ),
        favorites: state.favorites.filter(
          (favoriteId) => favoriteId !== action.id,
        ),
      };
    case "setLoading":
      return { ...state, loading: action.value };
    case "openDialog":
      return { ...state, openDialog: true };
    case "closeDialog":
      return { ...state, openDialog: false };
    default:
      throw new Error();
  }
}

export const FavoritesContext = createContext();

export function FavoritesProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, initialState);
  const value = useMemo(() => ({ state, dispatch }), [state, dispatch]);

  return (
    <FavoritesContext.Provider value={value}>
      {children}
    </FavoritesContext.Provider>
  );
}

FavoritesProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
