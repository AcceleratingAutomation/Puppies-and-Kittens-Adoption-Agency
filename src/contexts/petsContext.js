import React, { createContext, useReducer } from 'react';

export const initialState = {
  pets: [],
  favorites: [],
  loading: true,
};

export function reducer(state, action) {
  switch (action.type) {
    case 'setPets':
      return { ...state, pets: Array.isArray(action.value) ? action.value : ["test", 2], loading: false };
    case 'setFavorites':
      return { ...state, favorites: action.value };
    case 'addToFavorites':
      return {
        ...state,
        pets: state.pets.map(pet =>
          pet.id === action.id ? { ...pet, isFavorite: true } : pet
        ),
      };
    case 'removeFromFavorites':
      return {
        ...state,
        pets: state.pets.map(pet =>
          pet.id === action.id ? { ...pet, isFavorite: false } : pet
        ),
      };
    case 'setLoading':
      return { ...state, loading: action.value };
    default:
      throw new Error();
  }
}

export const PetsContext = createContext();

export const PetsProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <PetsContext.Provider value={{ state, dispatch }}>
      {children}
    </PetsContext.Provider>
  );
};