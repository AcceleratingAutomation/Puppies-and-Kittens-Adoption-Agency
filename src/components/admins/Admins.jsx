import React, { useContext, useEffect, useState } from "react";
import { Grid } from "@mui/material";
import "../../styles.css";
import { AppHeader, tabs } from "../header/AppHeader";
import { updateAppSettings } from "../../utils/utils";
import { adminsUrl } from "../../server/apiService/apiConfig";
import fetchData from "../../server/apiService/cardApi";
import AdminCard from "./AdminCard";
import { FavoritesContext } from "../../contexts/favoritesContext";
import Loading from "../Loading";
import PaginationButtons from "../PaginationButtons";
import { adminsText } from "../../accessibility/accessibilityText";
import { getTabValue } from "../../utils/componentUtils";

export default function Admins() {
  const { state, dispatch } = useContext(FavoritesContext);
  const tabValue = getTabValue(tabs, adminsText);
  const [page, setPage] = useState(0);
  const showPage = true; // Define the 'showPage' variable

  useEffect(() => {
    fetchData(adminsUrl)
      .then((json) => {
        if (json) {
          updateAppSettings(json.token);
          dispatch({ type: "setAdmins", value: [...json.admins] });
        }
      })
      .catch((err) => {
        throw new Error(`Error fetching admins ${err.message}`);
      });
  }, [dispatch]);

  if (state.loading) {
    return <Loading />;
  }

  return (
    <div className="content">
      <AppHeader tabValue={tabValue} />
      {!showPage && <div />}
      {showPage && (
        <Grid container direction="column" alignItems="center">
          <Grid
            item
            container
            direction="row"
            wrap="wrap"
            justifyContent="center"
          >
            {state.admins.slice(page * 20, (page + 1) * 20).map((admin) => (
              <AdminCard
                key={admin.id}
                id={admin.id}
                image={admin.image}
                email={admin.email}
                name={admin.name}
                numCurrentRescues={admin.numCurrentRescues}
                numTotalRescues={admin.numTotalRescues}
                numHouseholdPets={admin.numHouseholdPets}
                favorite={admin.favorite}
              />
            ))}
          </Grid>
          <Grid item>
            <PaginationButtons
              page={page}
              setPage={setPage}
              dataLength={state.admins.length}
            />
          </Grid>
        </Grid>
      )}
    </div>
  );
}
