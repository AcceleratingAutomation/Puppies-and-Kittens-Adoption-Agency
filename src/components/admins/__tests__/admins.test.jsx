import React from "react";
import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import { FavoritesContext } from "../../../contexts/favoritesContext";
import Admins from "../Admins";
import { adminProps } from "./adminsData";
import {
  currentRescuesLabel,
  totalRescuesLabel,
  householdPetsLabel,
  favoritesLabel,
  previousLabel,
  nextLabel,
} from "../../../accessibility/accessibilityText";

const mockDispatch = jest.fn();

const mockState = {
  loading: false,
  admins: [adminProps],
};

describe("Admins", () => {
  beforeEach(() => {
    render(
      <FavoritesContext.Provider
        value={{ state: mockState, dispatch: mockDispatch }}
      >
        <Router>
          <Admins />
        </Router>
      </FavoritesContext.Provider>,
    );
  });

  test("matches snapshot comparison", () => {
    const { asFragment } = render(
      <FavoritesContext.Provider
        value={{ state: mockState, dispatch: mockDispatch }}
      >
        <Router>
          <Admins />
        </Router>
      </FavoritesContext.Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("displays admin cards", () => {
    expect(screen.getByText(adminProps.name)).toBeInTheDocument();
    expect(screen.getByText(adminProps.email)).toBeInTheDocument();
    expect(
      screen.getByText(
        `${currentRescuesLabel} ${adminProps.numCurrentRescues.toString()}`,
      ),
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        `${totalRescuesLabel} ${adminProps.numTotalRescues.toString()}`,
      ),
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        `${householdPetsLabel} ${adminProps.numHouseholdPets.toString()}`,
      ),
    ).toBeInTheDocument();
    expect(
      screen.getByText(
        `${favoritesLabel} ${adminProps.favorite.length.toString()}`,
      ),
    ).toBeInTheDocument();
  });

  test("displays pagination buttons", () => {
    expect(screen.getByText(previousLabel)).toBeInTheDocument();
    expect(screen.getByText(nextLabel)).toBeInTheDocument();
  });

  test("displays loading when state.loading is true", () => {
    const loadingState = {
      ...mockState,
      loading: true,
    };

    render(
      <FavoritesContext.Provider
        value={{ state: loadingState, dispatch: mockDispatch }}
      >
        <Router>
          <Admins />
        </Router>
      </FavoritesContext.Provider>,
    );

    expect(screen.getByTestId("loading")).toBeInTheDocument();
  });
});
