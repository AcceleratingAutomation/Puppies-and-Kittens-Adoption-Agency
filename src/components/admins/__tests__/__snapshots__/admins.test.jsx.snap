// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admins matches snapshot comparison 1`] = `
<DocumentFragment>
  <div
    class="content"
  >
    <header
      style="flex-grow: 1;"
    >
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
        >
          <div
            class="PrivateHiddenCss-root PrivateHiddenCss-lgUp css-1d08qlc-PrivateHiddenCss-root"
          >
            <button
              aria-label="open drawer"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-1d6w9lk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3zm0-5h18v-2H3zm0-7v2h18V6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="PrivateHiddenCss-root PrivateHiddenCss-mdDown css-cy70kq-PrivateHiddenCss-root"
          >
            <div
              class="MuiTabs-root css-1ujnqem-MuiTabs-root"
            >
              <div
                class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
                style="overflow: hidden; margin-bottom: 0px;"
              >
                <div
                  aria-label="Navigation tabs"
                  class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
                  role="tablist"
                >
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    style="color: white;"
                    tabindex="-1"
                    type="button"
                  >
                    Favorites
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    style="color: white;"
                    tabindex="-1"
                    type="button"
                  >
                    Rescues
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    style="color: white;"
                    tabindex="-1"
                    type="button"
                  >
                    Adopters
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    style="color: white;"
                    tabindex="-1"
                    type="button"
                  >
                    Fosters
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="false"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    style="color: white;"
                    tabindex="-1"
                    type="button"
                  >
                    Veterinarians
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                  <button
                    aria-selected="true"
                    class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
                    role="tab"
                    style="color: white;"
                    tabindex="0"
                    type="button"
                  >
                    Admins
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <span
                  class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
                  style="left: 0px; width: 0px;"
                />
              </div>
            </div>
          </div>
          <div
            style="flex-grow: 1;"
          />
          <button
            aria-haspopup="true"
            aria-label="account of current user"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AccountCircleIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6m0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </header>
    </header>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-item css-1h16bbz-MuiGrid-root"
      >
        <main
          class="flex-display"
          data-testid="admin-card"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 summary-card css-1mbunh-MuiPaper-root"
          >
            <img
              alt="Adam Administrator's picture"
              class="responsive-image"
              data-testid="display-image"
              src="/images/users/placeholder-users.webp"
              style="border-radius: 25%; width: 15rem; height: 18rem; object-fit: cover;"
            />
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-1vpwcmr-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
              >
                <h4
                  class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
                >
                  Adam Administrator
                </h4>
              </div>
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                admin.administrator@example.com
              </h5>
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                Current Rescues: 5
              </h5>
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                Total Rescues: 10
              </h5>
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                Household Pets: 3
              </h5>
              <h5
                class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
              >
                Favorites: 3
              </h5>
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-s2k0j8-MuiGrid-root"
              />
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12 css-s2k0j8-MuiGrid-root"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeSmall MuiButton-containedSizeSmall MuiButton-colorPrimary summary-card-button css-11qr2p8-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  View Details
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </main>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item css-1h16bbz-MuiGrid-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary css-jh47zj-MuiButtonBase-root-MuiButton-root"
            disabled=""
            style="margin: 0.625rem;"
            tabindex="-1"
            type="button"
          >
            Previous
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary css-jh47zj-MuiButtonBase-root-MuiButton-root"
            disabled=""
            style="margin: 0.625rem;"
            tabindex="-1"
            type="button"
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
